builtin/ffi_overrides/voxelmanip.lua

```diff
@@ -0,0 +1,92 @@
+local ffi_shared = ...
+local ffi = ffi_shared.ffi
+local ie = ffi_shared.insecure_environment
+
+local metatable = ie.debug.getregistry().VoxelManip
+local C = ffi.C
+local rawget, rawset, type, tonumber, error, pcall =
+	_G.rawget, _G.rawset, _G.type, _G.tonumber, _G.error, _G.pcall
+local table_new = _G.table.new
+
+ffi.cdef[[
+struct __attribute__((aligned(__alignof__(uint32_t)))) MapNode {
+	uint16_t param0;
+	uint8_t param1, param2;
+};
+
+bool mtffi_vm_lock_area(void *ud);
+
+void mtffi_vm_unlock_area(void *ud);
+
+struct MapNode *mtffi_vm_get_data(void *ud);
+
+int32_t mtffi_vm_get_volume(void *ud);
+]]
+
+local check = ffi_shared.method_checker("VoxelManip", metatable)
+
+local function lock_area(self)
```

Comment:
Or actually, I just noticed we can remove the debug hook issues. Normally one wouldn't be able to check if there's a debug hook, because `debug.gethook` is a NYI and slow. But we can override `debug.sethook` to store in a lua table per coroutine thread if there's a hook. And if there's a hook on any coroutine you can fall back to calling the classic C-lua-interface function.
(I guess I've thrown away this idea at some point and forgot about it, because it's not possible in a mod.)

Reply:
Voxel manipulator locking can be removed if such an addition is merged.